File: resources/calc_server.erl
-module(calc_server).
-author("Duncan Paul Attard").

-include_lib("stdlib/include/assert.hrl").

-export([start/1]).
-export([loop/1]).

-spec start(N :: integer()) -> pid().
start(N) ->
    spawn(?MODULE, loop, [N]).

-spec loop(Tot :: integer()) -> no_return().
loop(Tot) ->
    receive
        {Clt, {add, A, B}} ->
            Clt ! {ok, A + B},
            loop(Tot + 1);
        {Clt, {mul, A, B}} ->
            Clt ! {ok, A * B},
            loop(Tot + 1);
        {Clt, stp} ->
            Clt ! {bye, Tot}
    end.

File: resources/calc_server.ex
defmodule Demo.CalcServer do
  # @spec start(n :: integer()) :: pid()
  def start(n) do
    spawn(__MODULE__, :loop, [n])
  end

  # @spec loop(tot :: integer()) :: no_return()
  def loop(tot) do
    receive do
      {clt, {:add, a, b}} ->
        send(clt, {:ok, a + b})
        loop(tot + 1)

      {clt, {:mul, a, b}} ->
        send(clt, {:ok, a * b})
        loop(tot + 1)

      {clt, :stp} ->
        send(clt, {:bye, tot})
    end
  end
end
